### DISCUSSION TITLE
raiseload('*') does not raise an Exception

### DISCUSSION THREAD

**Discussion by embecka** [sqlalchemy/sqlalchemy#8206] [Usage Questions] [answered]
Please, could somebody help me to understand why an exception is not raised when accessing the `user.addresses` attribute?

``` python
import logging
import sqlalchemy as sa
from sqlalchemy import orm

bind = sa.create_engine('sqlite://')
Session = orm.sessionmaker(bind=bind)
Base = orm.declarative_base()
logger = logging.getLogger('sqlalchemy.engine')


class User(Base):
    __tablename__ = "user"

    id = sa.Column(sa.Integer, primary_key=True)
    addresses = orm.relationship(lambda: Address, back_populates='user', cascade='all, delete-orphan')

    def __repr__(self):
        return f'User(id={self.id})'


class Address(Base):
    __tablename__ = "address"

    id = sa.Column(sa.Integer, primary_key=True)
    user_id = sa.Column(sa.Integer, sa.ForeignKey(User.id), nullable=False)
    user = orm.relationship(User, back_populates='addresses')

    def __repr__(self):
        return f'Address(id={self.id})'


def main():
    Base.metadata.create_all(bind, checkfirst=False)

    try:
        user = User(addresses=[Address()])

        with Session(expire_on_commit=False) as session, session.begin():
            session.add(user)

        level = logger.level
        logger.setLevel(logging.INFO)

        try:
            with Session() as session:
                with session.begin():
                    users = session.query(User).options(orm.raiseload('*')).all()
                    print(users)

                with session.begin():
                    user = session.get(User, user.id, options=[orm.raiseload('*')])
                    print(user.addresses)
        finally:
            logger.setLevel(level)
    finally:
        Base.metadata.drop_all(bind, checkfirst=False)


if __name__ == "__main__":
    logging.basicConfig()
    main()
```

**Comment by zzzeek** [comment, marked as answer]
hi -

thanks for the very clear example.  you've found a specific combination of things where everything is doing what it should, but the end result looks odd.

The main thing happening here is that you cannot rely upon ``Session.get()`` to change the object returned in any way if the object might be already present in the identity map - ``Session.get()`` gives you the existing object as is and does not apply ``raiseload()`` because the User object was already loaded from the previous operation.      To force ``Session.get()`` to load the object and apply new options, use ``populate_existing=True``::

```py
user = session.get(User, user.id, options=[orm.raiseload('*')], populate_existing=True)
```

then, the first ``raiseload`` was cancelled because the ``User`` object was expired.    

Generally I would turn off ``expire_on_commit`` here if you are running multiple transactions and are attempting to re-use the results from the previous transaction in the new one:

```py
with Session(expire_on_commit=False) as session:
```

**Reply by embecka** [reply]
I wasn't aware that loading strategies expire with the object, this is the only thing that seems odd to me, but I accept it. The `populate_existing` argument really works, I didn't realize it could also be used to solve my problem. The generated SQL statements are the same, but now an exception is thrown. Thanks for the explanation and the work on `sqlalchemy`.
